---
title: "CUT&TAG H3K27ac antibody comparisons (no duplicates)"
author: "Leyla Abbasova"
output:
  rmarkdown::html_document:
    theme: default
    highlight: tango
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    number_sections: no
    self_contained: yes
    df_print: paged
---

```{r setup, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(stringr)
library(ggplot2)
library(ggpubr)
library(viridis)
library(GenomicRanges)
library(IRanges)
library(chromVAR)
library(ChIPseeker)
library(ReactomePA)
library(rtracklayer)
library(clusterProfiler)
library(rGADEM)
library(BRGenomics)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
library(GenometriCorr)
```


# General

## Alignment summary 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8, out.width = "100%"}
datadir="/Volumes/la420/home/CUTnTAG/antibodies/bowtie2_summary/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

## collect the alignment results from the bowtie2 alignment summary files
alignResult = c()
for(sample in sampleList){
  alignRes = read.table(paste0(datadir, sample, "_trimmed_bowtie2.txt"), header = FALSE, fill = TRUE)
  alignRate = substr(alignRes$V1[6], 1, nchar(as.character(alignRes$V1[6]))-1)
  alignResult = data.frame(Antibody = sample, SequencingDepth = alignRes$V1[1] %>% as.character %>% as.numeric, 
                           MappedFragNum_hg19 = alignRes$V1[4] %>% as.character %>% as.numeric + alignRes$V1[5] %>% as.character %>% as.numeric,
                           AlignmentRate_hg19 = alignRate %>% as.numeric)  %>% rbind(alignResult, .)
}
alignResult %>% mutate(AlignmentRate_hg19 = paste0(AlignmentRate_hg19, "%"))
```


## Duplicates

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
## summarize the duplication information from the picard summary outputs
picarddir = "/Volumes/la420/home/CUTnTAG/antibodies/picard_summary/"

dupResult = c()
for(sample in sampleList){
  dupRes = read.table(paste0(picarddir, sample, "_picard.rmDup.txt"), header = TRUE, fill = TRUE)
  dupResult = data.frame(Antibody = sample, MappedFragNum_hg19 = dupRes$READ_PAIRS_EXAMINED[1] %>% as.character %>% as.numeric, DuplicationRate = dupRes$PERCENT_DUPLICATION[1] %>% as.character %>% as.numeric * 100, EstimatedLibrarySize = dupRes$ESTIMATED_LIBRARY_SIZE[1] %>% as.character %>% as.numeric) %>% mutate(UniqueFragNum = MappedFragNum_hg19 * (1-DuplicationRate/100))  %>% rbind(dupResult, .)
}

alignDupSummary = left_join(alignResult, dupResult, by = c("Antibody", "MappedFragNum_hg19")) %>% mutate(DuplicationRate = paste0(DuplicationRate, "%"))
alignDupSummary
```


## Fragment length    {.tabset}

### Without duplicates 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
## collect the fragment size information
fragdir = "/Volumes/la420/home/CUTnTAG/antibodies/fragmentLen/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

fragLen = c()
for(sample in sampleList){
  fragLen = read.table(paste0(fragdir, sample, "_rmDup_fragmentLen.txt"), header = FALSE) %>% mutate(fragLen = V1 %>% as.numeric, fragCount = V2 %>% as.numeric, Weight = as.numeric(V2)/sum(as.numeric(V2)), Antibody = sample) %>% rbind(fragLen, .) 
}

fig5A = fragLen %>% ggplot(aes(x = Antibody, y = fragLen, weight = Weight, fill = Antibody)) +
    geom_violin(bw = 5) +
    scale_y_continuous(breaks = seq(0, 800, 50)) +
    scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "magma", alpha = 0.8) +
    scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +
    theme_bw(base_size = 20) +
    ggpubr::rotate_x_text(angle = 20) +
    ylab("Fragment Length") +
    xlab("")

fig5B = fragLen %>% ggplot(aes(x = fragLen, y = fragCount, color = Antibody, group = Antibody)) +
  geom_line(size = 1) +
  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "magma") +
  theme_bw(base_size = 20) +
  xlab("Fragment Length") +
  ylab("Count") +
  coord_cartesian(xlim = c(0, 500))

ggarrange(fig5A, fig5B, ncol = 2)
```


### With duplicates 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
## collect the fragment size information
fragdir = "/Volumes/la420/home/CUTnTAG/antibodies_withDup/fragmentLen/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

fragLen = c()
for(sample in sampleList){
  fragLen = read.table(paste0(fragdir, sample, "_withDup_fragmentLen.txt"), header = FALSE) %>% mutate(fragLen = V1 %>% as.numeric, fragCount = V2 %>% as.numeric, Weight = as.numeric(V2)/sum(as.numeric(V2)), Antibody = sample) %>% rbind(fragLen, .) 
}

fig5A = fragLen %>% ggplot(aes(x = Antibody, y = fragLen, weight = Weight, fill = Antibody)) +
    geom_violin(bw = 5) +
    scale_y_continuous(breaks = seq(0, 800, 50)) +
    scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "viridis", alpha = 0.8) +
    scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +
    theme_bw(base_size = 20) +
    ggpubr::rotate_x_text(angle = 20) +
    ylab("Fragment Length") +
    xlab("")

fig5B = fragLen %>% ggplot(aes(x = fragLen, y = fragCount, color = Antibody, group = Antibody)) +
  geom_line(size = 1) +
  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "viridis") +
  theme_bw(base_size = 20) +
  xlab("Fragment Length") +
  ylab("Count") +
  coord_cartesian(xlim = c(0, 500))

ggarrange(fig5A, fig5B, ncol = 2)
```


# SEACR

## 1. Import peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir_S = "/Volumes/la420/home/CUTnTAG/antibodies/SEACR"

Diagenode_50x_path = file.path(peakdir_S, "Diagenode_50x_seacr_top0.01.peaks.stringent.bed")
Diagenode_50x = ChIPseeker::readPeakFile(Diagenode_50x_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Abcam_ab4729_path = file.path(peakdir_S, "Abcam-ab4729_seacr_top0.01.peaks.stringent.bed")
Abcam_ab4729 = ChIPseeker::readPeakFile(Abcam_ab4729_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Diagenode_100x_path = file.path(peakdir_S, "Diagenode_100x_seacr_top0.01.peaks.stringent.bed")
Diagenode_100x = ChIPseeker::readPeakFile(Diagenode_100x_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

ActiveMotif_path = file.path(peakdir_S, "ActiveMotif_seacr_top0.01.peaks.stringent.bed")
ActiveMotif = ChIPseeker::readPeakFile(ActiveMotif_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Abcam_ab177178_path = file.path(peakdir_S, "Abcam-ab177178_seacr_top0.01.peaks.stringent.bed")
Abcam_ab177178 = ChIPseeker::readPeakFile(Abcam_ab177178_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)
```


## 2. Peak information

### Peak numbers

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir_S = "/Volumes/la420/home/CUTnTAG/antibodies/SEACR/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

Total_CT_peaks = c()
for(sample in sampleList){
    peakInfo_S = read.table(paste0(peakdir_S, sample, "_seacr_top0.01.peaks.stringent.bed"), header = FALSE, fill = TRUE)
    Total_CT_peaks = data.frame(Total_CT_peaks = nrow(peakInfo_S), Antibody = sample) %>% rbind(Total_CT_peaks, .)
 }

Total_CT_peaks %>% dplyr::select(Antibody, Total_CT_peaks)
```


#### Peak and sequencing/alignment results

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
PeaksAlignmentSummary = left_join(Total_CT_peaks, alignDupSummary, by = "Antibody")
PeaksAlignmentSummary = PeaksAlignmentSummary[c("Antibody", "SequencingDepth", "MappedFragNum_hg19", "UniqueFragNum", "DuplicationRate", "Total_CT_peaks")]
PeaksAlignmentSummary
```


```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
ggplot(PeaksAlignmentSummary, aes(x = SequencingDepth, y = Total_CT_peaks)) + geom_point(aes(size = DuplicationRate)) + geom_text(aes(label = Antibody), nudge_y = 500, hjust = "inward")
```


### Peak widths

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
samples = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)

for(sample in samples){
  print(GenomicRanges::width(sample) %>% summary())
}
```


### Fraction of reads in peaks
* ENCODE ChIP-seq FRiP score is ~44% (https://www.encodeproject.org/files/ENCFF044JNJ/)
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
datadir="/Volumes/la420/home/CUTnTAG/antibodies"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

inPeakData = c()
for(sample in sampleList){
  peakRes = read.table(paste0(datadir, "/SEACR/", sample, "_seacr_top0.01.peaks.stringent.bed"), header = FALSE, fill = TRUE)
  peak.gr = GenomicRanges::GRanges(seqnames = peakRes$V1, IRanges(start = peakRes$V2, end = peakRes$V3), strand = "*")
  bamFile = paste0(datadir, "/bam/", sample, "_bowtie2_rmDup.mapped.bam")
  fragment_counts = getCounts(bamFile, peak.gr, paired = TRUE, by_rg = FALSE, format = "bam")
  inPeakN = counts(fragment_counts)[,1] %>% sum
  inPeakData = rbind(inPeakData, data.frame(inPeakN = inPeakN, Antibody = sample))
}

frip = left_join(alignDupSummary, inPeakData, by = "Antibody") %>% mutate(frip = inPeakN/UniqueFragNum * 100)
frip %>% dplyr::select(Antibody, SequencingDepth, MappedFragNum_hg19, AlignmentRate_hg19, UniqueFragNum, FragInPeakNum = inPeakN, FRiPs = frip)
```


## 3. Coverage
Remove largest peaks (in all cases, under 30 peaks per sample):
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=8}
ActiveMotif_adj = data.frame(ActiveMotif)
ActiveMotif_adj = ActiveMotif_adj[ActiveMotif_adj$V4 < 50000, ] %>% GenomicRanges::GRanges()

Diagenode_100x_adj = data.frame(Diagenode_100x)
Diagenode_100x_adj = Diagenode_100x_adj[Diagenode_100x_adj$V4 < 100000, ] %>% GenomicRanges::GRanges()

Diagenode_50x_adj = data.frame(Diagenode_50x)
Diagenode_50x_adj = Diagenode_50x_adj[Diagenode_50x_adj$V4 < 400000, ] %>% GenomicRanges::GRanges()

Abcam_ab4729_adj = data.frame(Abcam_ab4729)
Abcam_ab4729_adj = Abcam_ab4729_adj[Abcam_ab4729_adj$V4 < 400000, ] %>% GenomicRanges::GRanges()

Abcam_ab177178_adj = data.frame(Abcam_ab177178)
Abcam_ab177178_adj = Abcam_ab177178_adj[Abcam_ab177178_adj$V4 < 100000, ] %>% GenomicRanges::GRanges()
```


### Covplots    {.tabset}

#### Active Motif

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = ActiveMotif_adj,
                    weightCol = "V4",
                    title = "CUT&TAG ActiveMotif (1:100) H3K27ac peaks (SEACR)")
```

#### Diagenode (1:100)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Diagenode_100x_adj,
                    weightCol = "V4",
                    title = "CUT&TAG Diagenode (1:100) H3K27ac peaks (SEACR)")
```

#### Diagenode (1:50)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Diagenode_50x_adj,
                    weightCol = "V4",
                    title = "CUT&TAG Diagenode (1:50) H3K27ac peaks (SEACR)")
```

#### Abcam-ab4729

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Abcam_ab4729_adj,
                    weightCol = "V4",
                    title = "CUT&TAG Abcam-ab4729 (1:100) H3K27ac peaks (SEACR)")
```

#### Abcam-ab177178

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Abcam_ab177178_adj,
                    weightCol = "V4",
                    title = "CUT&TAG Abcam-ab177178 (1:100) H3K27ac peaks (SEACR)")
```


## 4. ENCODE peak overlap
Using ENCODE narrowPeak file with replicated H3K27ac peaks in K562 (ENCFF044JNJ)
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
extraCols_narrowPeak = c(signalValue = "numeric", pValue = "numeric", qValue = "numeric", peak = "integer")

ENCODE_narrowPeaks = rtracklayer::import("https://www.encodeproject.org/files/ENCFF044JNJ/@@download/ENCFF044JNJ.bed.gz", format = "BED", extraCols = extraCols_narrowPeak)
```


### Count total C&T peaks falling into ENCODE
Discard values of 0 (corresponding to peak ranges with 0 ENCODE overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peak_list = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)
total_overlapping_peaks_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = peak_list[[i]], subject = ENCODE_narrowPeaks)
  CT_peaks_in_ENCODE = overlaps[overlaps != 0] %>% length()
  total_overlapping_peaks_list = c(total_overlapping_peaks_list, CT_peaks_in_ENCODE)
}

Total_CT_peaks_falling_into_ENCODE = data.frame(Antibody = sampleList, Total_CT_peaks_falling_into_ENCODE = total_overlapping_peaks_list)
Total_CT_peaks_falling_into_ENCODE
```


### Count total ENCODE peaks falling into C&T
Discard values of 0 (corresponding to peak ranges with 0 C&T overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peak_list = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)
captured_ENCODE_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = ENCODE_narrowPeaks, subject = peak_list[[i]])
  ENCODE_captured_peaks = overlaps[overlaps != 0] %>% length()
  captured_ENCODE_list = c(captured_ENCODE_list, ENCODE_captured_peaks)
}

Total_captured_ENCODE_peaks = data.frame(Antibody = sampleList, Total_captured_ENCODE_peaks = captured_ENCODE_list)
Total_captured_ENCODE_peaks
```


### Summarize

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ENCODE_overlaps = Total_captured_ENCODE_peaks %>% left_join(Total_CT_peaks, by = "Antibody") %>% left_join(Total_CT_peaks_falling_into_ENCODE, by = "Antibody")

ENCODE_overlaps$Percentage_of_CT_peaks_falling_into_ENCODE = ENCODE_overlaps$Total_CT_peaks_falling_into_ENCODE / ENCODE_overlaps$Total_CT_peaks * 100
ENCODE_overlaps$Percentage_of_total_ENCODE = ENCODE_overlaps$Total_captured_ENCODE_peaks / length(ENCODE_narrowPeaks) * 100

ENCODE_overlaps = ENCODE_overlaps[c("Antibody", "Total_CT_peaks", "Total_CT_peaks_falling_into_ENCODE", "Percentage_of_CT_peaks_falling_into_ENCODE", "Total_captured_ENCODE_peaks", "Percentage_of_total_ENCODE")]
ENCODE_overlaps
```


### ENCODE peaks captured by C&T (chart)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=4}
ENCODE_overlaps$Antibody = factor(ENCODE_overlaps$Antibody, levels = ENCODE_overlaps$Antibody[order(ENCODE_overlaps$Percentage_of_total_ENCODE)])

#x_labels = c("Active Motif (1:100)", "Diagenode (1:100)", "Diagenode (1:50)", "Abcam 177178 (1:100)", "Abcam 4729 (1:100)")

ENCODE_overlap_plot = ggplot(ENCODE_overlaps) +
  geom_col(aes(x = Antibody, y = 100), fill = "grey90") +
  geom_col(aes(x = reorder(Antibody, Percentage_of_total_ENCODE), y = Percentage_of_total_ENCODE, fill = Antibody), show.legend = FALSE) +
  ylim(0, 100) +
  ylab("% of ENCODE peaks captured") +
  xlab("Antibody") +
  #scale_x_discrete(labels = x_labels) +
  ggtitle("ENCODE peaks captured by C&T") +
  theme_bw()
  
ENCODE_overlap_plot
```


#### Visualize overlaps   {.tabset}

##### Abcam-ab4729

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
Abcam_ab4729_df = data.frame(Abcam_ab4729)
Abcam_ab4729_IRanges = IRanges(start = Abcam_ab4729_df$start, Abcam_ab4729_df$end)

ENCODE_narrowPeaks_df = data.frame(ENCODE_narrowPeaks)
ENCODE_narrowPeaks_IRanges = IRanges(start = ENCODE_narrowPeaks_df$start, ENCODE_narrowPeaks_df$end)

GenometriCorr::VisualiseTwoIRanges(Abcam_ab4729_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Abcam_ab4729", nameB = "ENCODE_narrowPeaks")
```

##### Diagenode_50x

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
Diagenode_50x_df = data.frame(Diagenode_50x)
Diagenode_50x_IRanges = IRanges(start = Diagenode_50x_df$start, Diagenode_50x_df$end)

GenometriCorr::VisualiseTwoIRanges(Diagenode_50x_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Diagenode_50x", nameB = "ENCODE_narrowPeaks")
```

##### ActiveMotif

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
ActiveMotif_df = data.frame(ActiveMotif)
ActiveMotif_IRanges = IRanges(start = ActiveMotif_df$start, ActiveMotif_df$end)

GenometriCorr::VisualiseTwoIRanges(ActiveMotif_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Diagenode_50x", nameB = "ENCODE_narrowPeaks")
```


### Proportion of C&T peaks falling into ENCODE (chart)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=4}
ENCODE_overlaps$Antibody = factor(ENCODE_overlaps$Antibody, levels = ENCODE_overlaps$Antibody[order(ENCODE_overlaps$Percentage_of_total_ENCODE)])

#x_labels = c("Active Motif (1:100)", "Diagenode (1:100)", "Diagenode (1:50)", "Abcam 177178 (1:100)", "Abcam 4729 (1:100)")

Peaks_in_ENCODE_plot = ggplot(ENCODE_overlaps) +
  geom_col(aes(x = Antibody, y = 100), fill = "grey90") +
  geom_col(aes(x = reorder(Antibody, Percentage_of_total_ENCODE), y = Percentage_of_CT_peaks_falling_into_ENCODE, fill = Antibody), show.legend = FALSE) +
  ylim(0, 100) +
  ylab("% of sample peaks falling into ENCODE") +
  xlab("Antibody") +
  #scale_x_discrete(labels = x_labels) +
  ggtitle("Proportion of C&T peaks falling into ENCODE") +
  theme_bw()
  
Peaks_in_ENCODE_plot
```


## 5. Average TSS peak profiles

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
promoters = ChIPseeker::getPromoters(TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, upstream = 3000, downstream = 3000)
peak_files = list("Active Motif (1:100)" = ActiveMotif_path,
                  "Diagenode (1:100)" = Diagenode_100x_path,
                  "Diagenode (1:50)" = Diagenode_50x_path,
                  "Abcam 177178 (1:100)" = Abcam_ab177178_path,
                  "Abcam 4729 (1:100)" = Abcam_ab4729_path,
                  "ENCODE narrow peaks" = ENCODE_narrowPeaks)

tagMatrixList = lapply(peak_files, getTagMatrix, windows = promoters)

ChIPseeker::plotAvgProf(tagMatrixList, xlim = c(-3000, 3000), conf = 0.95, resample = 500, facet = "row")
```


## 6. Peak heatmaps

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
tagHeatmap(tagMatrixList, xlim = c(-3000, 3000), color = NULL)
```


## 7. Annotation

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakAnnoList = lapply(peak_files, annotatePeak, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, tssRegion = c(-3000, 3000), verbose = FALSE)

ChIPseeker::plotAnnoBar(peakAnnoList)
```


## 8. Distance to TSS

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ChIPseeker::plotDistToTSS(peakAnnoList)
```


## 9. Enrichment analysis

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=13, fig.height=8}
genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))
compKEGG = clusterProfiler::compareCluster(geneCluster = genes,
                                           fun = "enrichKEGG",
                                           pvalueCutoff = 0.05,
                                           pAdjustMethod = "BH")
clusterProfiler::dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


## 10. Gene overlaps

Without Active Motif
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8}
peaks = list("Diagenode (1:100)" = Diagenode_100x,
             "Diagenode (1:50)" = Diagenode_50x,
             "Abcam 177178 (1:100)" = Abcam_ab177178,
             "Abcam 4729 (1:100)" = Abcam_ab4729)

peakAnnoList = lapply(peaks, annotatePeak, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, tssRegion = c(-3000, 3000), verbose = FALSE)

genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
vennplot(genes)
```


## 11. Peak overlaps

Without Active Motif
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8}
peaks = list("Diagenode (1:100)" = GenomicRanges::GRanges(Diagenode_100x),
             "Diagenode (1:50)" = GenomicRanges::GRanges(Diagenode_50x),
             "Abcam 177178 (1:100)" = GenomicRanges::GRanges(Abcam_ab177178),
             "Abcam 4729 (1:100)" = GenomicRanges::GRanges(Abcam_ab4729))

vennplot(peaks)
```


## 12. Motif enrichment

### HOMER

```{r, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=5, fig.height=10}
motifdir = "/Volumes/la420/home/CUTnTAG/antibodies/SEACR/motifs/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

for(sample in sampleList){
  motifs = read.csv(paste0(motifdir, sample, "/knownResults.txt"), sep = "\t", header = TRUE) %>% data.frame()
  print(head(motifs))
}
```



# MACS2

## 1. Import peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir_M = "/Volumes/la420/home/CUTnTAG/antibodies/MACS2"

Diagenode_50x_path = file.path(peakdir_M, "Diagenode_50x_q0.05_peaks.narrowPeak")
Diagenode_50x = ChIPseeker::readPeakFile(Diagenode_50x_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Abcam_ab4729_path = file.path(peakdir_M, "Abcam-ab4729_q0.05_peaks.narrowPeak")
Abcam_ab4729 = ChIPseeker::readPeakFile(Abcam_ab4729_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Diagenode_100x_path = file.path(peakdir_M, "Diagenode_100x_q0.05_peaks.narrowPeak")
Diagenode_100x = ChIPseeker::readPeakFile(Diagenode_100x_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

ActiveMotif_path = file.path(peakdir_M, "ActiveMotif_q0.05_peaks.narrowPeak")
ActiveMotif = ChIPseeker::readPeakFile(ActiveMotif_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

Abcam_ab177178_path = file.path(peakdir_M, "Abcam-ab177178_q0.05_peaks.narrowPeak")
Abcam_ab177178 = ChIPseeker::readPeakFile(Abcam_ab177178_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)
```


## 2. Peak information

### Peak numbers

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir_M = "/Volumes/la420/home/CUTnTAG/antibodies/MACS2/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

Total_CT_peaks = c()
for(sample in sampleList){
    peakInfo_M = read.table(paste0(peakdir_M, sample, "_q0.05_peaks.narrowPeak"), header = FALSE, fill = TRUE) 
    Total_CT_peaks = data.frame(Total_CT_peaks = nrow(peakInfo_M), Antibody = sample) %>% rbind(Total_CT_peaks, .)
 }

Total_CT_peaks %>% dplyr::select(Antibody, Total_CT_peaks)
```


#### Peak and sequencing/alignment results

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
PeaksAlignmentSummary = left_join(Total_CT_peaks, alignDupSummary, by = "Antibody")
PeaksAlignmentSummary = PeaksAlignmentSummary[c("Antibody", "SequencingDepth", "MappedFragNum_hg19", "UniqueFragNum", "DuplicationRate", "Total_CT_peaks")]
PeaksAlignmentSummary
```


```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
ggplot(PeaksAlignmentSummary, aes(x = SequencingDepth, y = Total_CT_peaks)) + geom_point(aes(size = DuplicationRate)) + geom_text(aes(label = Antibody), nudge_y = 500, hjust = "inward")
```


### Peak widths

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
samples = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)

for(sample in samples){
  print(GenomicRanges::width(sample) %>% summary())
}
```


### Fraction of reads in peaks
* ENCODE ChIP-seq FRiP score is ~44% (https://www.encodeproject.org/files/ENCFF044JNJ/)
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
datadir="/Volumes/la420/home/CUTnTAG/antibodies"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

inPeakData = c()
for(sample in sampleList){
  peakRes = read.table(paste0(datadir, "/MACS2/", sample, "_q0.05_peaks.narrowPeak"), header = FALSE, fill = TRUE)
  peak.gr = GenomicRanges::GRanges(seqnames = peakRes$V1, IRanges(start = peakRes$V2, end = peakRes$V3), strand = "*")
  bamFile = paste0(datadir, "/bam/", sample, "_bowtie2_rmDup.mapped.bam")
  fragment_counts = getCounts(bamFile, peak.gr, paired = TRUE, by_rg = FALSE, format = "bam")
  inPeakN = counts(fragment_counts)[,1] %>% sum
  inPeakData = rbind(inPeakData, data.frame(inPeakN = inPeakN, Antibody = sample))
}

frip = left_join(alignDupSummary, inPeakData, by = "Antibody") %>% mutate(frip = inPeakN/UniqueFragNum * 100)
frip %>% dplyr::select(Antibody, SequencingDepth, MappedFragNum_hg19, AlignmentRate_hg19, UniqueFragNum, FragInPeakNum = inPeakN, FRiPs = frip)
```


## 3. Coverage
Remove largest peaks (in all cases, under 50 peaks per sample):
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=6, fig.height=8}
ActiveMotif_adj = data.frame(ActiveMotif)
ActiveMotif_adj = ActiveMotif_adj[ActiveMotif_adj$V5 < 1500, ] %>% GenomicRanges::GRanges()

Diagenode_100x_adj = data.frame(Diagenode_100x)
Diagenode_100x_adj = Diagenode_100x_adj[Diagenode_100x_adj$V5 < 1500, ] %>% GenomicRanges::GRanges()

Diagenode_50x_adj = data.frame(Diagenode_50x)
Diagenode_50x_adj = Diagenode_50x_adj[Diagenode_50x_adj$V5 < 1500, ] %>% GenomicRanges::GRanges()

Abcam_ab4729_adj = data.frame(Abcam_ab4729)
Abcam_ab4729_adj = Abcam_ab4729_adj[Abcam_ab4729_adj$V5 < 1500, ] %>% GenomicRanges::GRanges()

Abcam_ab177178_adj = data.frame(Abcam_ab177178)
Abcam_ab177178_adj = Abcam_ab177178_adj[Abcam_ab177178_adj$V5 < 1500, ] %>% GenomicRanges::GRanges()
```


### Covplots    {.tabset}

#### Active Motif

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = ActiveMotif_adj,
                    weightCol = "V5",
                    title = "CUT&TAG ActiveMotif (1:100) H3K27ac peaks (MACS2)")
```

#### Diagenode (1:100)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Diagenode_100x_adj,
                    weightCol = "V5",
                    title = "CUT&TAG Diagenode (1:100) H3K27ac peaks (MACS2)")
```

#### Diagenode (1:50)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Diagenode_50x_adj,
                    weightCol = "V5",
                    title = "CUT&TAG Diagenode (1:50) H3K27ac peaks (MACS2)")
```

#### Abcam-ab4729

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Abcam_ab4729_adj,
                    weightCol = "V5",
                    title = "CUT&TAG Abcam-ab4729 (1:100) H3K27ac peaks (MACS2)")
```

#### Abcam-ab177178

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
ChIPseeker::covplot(peak = Abcam_ab177178_adj,
                    weightCol = "V5",
                    title = "CUT&TAG Abcam-ab177178 (1:100) H3K27ac peaks (MACS2)")
```


## 4. ENCODE peak overlap

### Count total C&T peaks falling into ENCODE
Discard values of 0 (corresponding to peak ranges with 0 ENCODE overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peak_list = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)
total_overlapping_peaks_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = peak_list[[i]], subject = ENCODE_narrowPeaks)
  CT_peaks_in_ENCODE = overlaps[overlaps != 0] %>% length()
  total_overlapping_peaks_list = c(total_overlapping_peaks_list, CT_peaks_in_ENCODE)
}

Total_CT_peaks_falling_into_ENCODE = data.frame(Antibody = sampleList, Total_CT_peaks_falling_into_ENCODE = total_overlapping_peaks_list)
Total_CT_peaks_falling_into_ENCODE
```


### Count total ENCODE peaks falling into C&T
Discard values of 0 (corresponding to peak ranges with 0 C&T overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peak_list = list(ActiveMotif, Diagenode_100x, Diagenode_50x, Abcam_ab177178, Abcam_ab4729)
captured_ENCODE_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = ENCODE_narrowPeaks, subject = peak_list[[i]])
  ENCODE_captured_peaks = overlaps[overlaps != 0] %>% length()
  captured_ENCODE_list = c(captured_ENCODE_list, ENCODE_captured_peaks)
}

Total_captured_ENCODE_peaks = data.frame(Antibody = sampleList, Total_captured_ENCODE_peaks = captured_ENCODE_list)
Total_captured_ENCODE_peaks
```


### Summarize

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ENCODE_overlaps = Total_captured_ENCODE_peaks %>% left_join(Total_CT_peaks, by = "Antibody") %>% left_join(Total_CT_peaks_falling_into_ENCODE, by = "Antibody")

ENCODE_overlaps$Percentage_of_CT_peaks_falling_into_ENCODE = ENCODE_overlaps$Total_CT_peaks_falling_into_ENCODE / ENCODE_overlaps$Total_CT_peaks * 100
ENCODE_overlaps$Percentage_of_total_ENCODE = ENCODE_overlaps$Total_captured_ENCODE_peaks / length(ENCODE_narrowPeaks) * 100

ENCODE_overlaps = ENCODE_overlaps[c("Antibody", "Total_CT_peaks", "Total_CT_peaks_falling_into_ENCODE", "Percentage_of_CT_peaks_falling_into_ENCODE", "Total_captured_ENCODE_peaks", "Percentage_of_total_ENCODE")]
ENCODE_overlaps
```


### ENCODE peaks captured by C&T (chart)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=4}
ENCODE_overlaps$Antibody = factor(ENCODE_overlaps$Antibody, levels = ENCODE_overlaps$Antibody[order(ENCODE_overlaps$Percentage_of_total_ENCODE)])

#x_labels = c("Active Motif (1:100)", "Abcam 177178 (1:100)", "Diagenode (1:100)", "Diagenode (1:50)", "Abcam 4729 (1:100)")

ENCODE_overlap_plot = ggplot(ENCODE_overlaps) +
  geom_col(aes(x = Antibody, y = 100), fill = "grey90") +
  geom_col(aes(x = reorder(Antibody, Percentage_of_total_ENCODE), y = Percentage_of_total_ENCODE, fill = Antibody), show.legend = FALSE) +
  ylim(0, 100) +
  ylab("% of ENCODE peaks captured") +
  xlab("Antibody") +
  #scale_x_discrete(labels = x_labels) +
  ggtitle("ENCODE peaks captured by C&T") +
  theme_bw()
  
ENCODE_overlap_plot
```


#### Visualize overlaps   {.tabset}

##### Abcam-ab4729

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
Abcam_ab4729_df = data.frame(Abcam_ab4729)
Abcam_ab4729_IRanges = IRanges(start = Abcam_ab4729_df$start, Abcam_ab4729_df$end)

ENCODE_narrowPeaks_IRanges = IRanges(start = ENCODE_narrowPeaks_df$start, ENCODE_narrowPeaks_df$end)

GenometriCorr::VisualiseTwoIRanges(Abcam_ab4729_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Abcam_ab4729", nameB = "ENCODE_narrowPeaks")
```

##### Diagenode_50x

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
Diagenode_50x_df = data.frame(Diagenode_50x)
Diagenode_50x_IRanges = IRanges(start = Diagenode_50x_df$start, Diagenode_50x_df$end)

GenometriCorr::VisualiseTwoIRanges(Diagenode_50x_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Diagenode_50x", nameB = "ENCODE_narrowPeaks")
```

##### ActiveMotif

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=5}
ActiveMotif_df = data.frame(ActiveMotif)
ActiveMotif_IRanges = IRanges(start = ActiveMotif_df$start, ActiveMotif_df$end)

GenometriCorr::VisualiseTwoIRanges(ActiveMotif_IRanges, ENCODE_narrowPeaks_IRanges, nameA = "Diagenode_50x", nameB = "ENCODE_narrowPeaks")
```


### Proportion of C&T peaks falling into ENCODE (chart)

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=7, fig.height=4}
ENCODE_overlaps$Antibody = factor(ENCODE_overlaps$Antibody, levels = ENCODE_overlaps$Antibody[order(ENCODE_overlaps$Percentage_of_total_ENCODE)])

#x_labels = c("Active Motif (1:100)", "Diagenode (1:100)", "Diagenode (1:50)", "Abcam 177178 (1:100)", "Abcam 4729 (1:100)")

Peaks_in_ENCODE_plot = ggplot(ENCODE_overlaps) +
  geom_col(aes(x = Antibody, y = 100), fill = "grey90") +
  geom_col(aes(x = reorder(Antibody, Percentage_of_total_ENCODE), y = Percentage_of_CT_peaks_falling_into_ENCODE, fill = Antibody), show.legend = FALSE) +
  ylim(0, 100) +
  ylab("% of sample peaks falling into ENCODE") +
  xlab("Antibody") +
  #scale_x_discrete(labels = x_labels) +
  ggtitle("Proportion of C&T peaks falling into ENCODE") +
  theme_bw()
  
Peaks_in_ENCODE_plot
```


## 5. Average TSS peak profiles

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
promoters = ChIPseeker::getPromoters(TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, upstream = 3000, downstream = 3000)
peak_files = list("Active Motif (1:100)" = ActiveMotif_path,
                  "Diagenode (1:100)" = Diagenode_100x_path,
                  "Diagenode (1:50)" = Diagenode_50x_path,
                  "Abcam 177178 (1:100)" = Abcam_ab177178_path,
                  "Abcam 4729 (1:100)" = Abcam_ab4729_path,
                  "ENCODE narrow peaks" = ENCODE_narrowPeaks)

tagMatrixList = lapply(peak_files, getTagMatrix, windows = promoters)

ChIPseeker::plotAvgProf(tagMatrixList, xlim = c(-3000, 3000), conf = 0.95, resample = 500, facet = "row")
```


## 6. Peak heatmaps

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
tagHeatmap(tagMatrixList, xlim = c(-3000, 3000), color = NULL)
```


## 7. Annotation

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakAnnoList = lapply(peak_files, annotatePeak, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, tssRegion = c(-3000, 3000), verbose = FALSE)

ChIPseeker::plotAnnoBar(peakAnnoList)
```


## 8. Distance to TSS

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ChIPseeker::plotDistToTSS(peakAnnoList)
```


## 9. Enrichment analysis

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=13, fig.height=8}
genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))
compKEGG = clusterProfiler::compareCluster(geneCluster = genes,
                                           fun = "enrichKEGG",
                                           pvalueCutoff = 0.05,
                                           pAdjustMethod = "BH")
clusterProfiler::dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


## 10. Gene overlaps

Without Active Motif
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8}
peak_files = list("Diagenode (1:100)" = Diagenode_100x,
                  "Diagenode (1:50)" = Diagenode_50x,
                  "Abcam 177178 (1:100)" = Abcam_ab177178,
                  "Abcam 4729 (1:100)" = Abcam_ab4729)

peakAnnoList = lapply(peak_files, annotatePeak, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, tssRegion = c(-3000, 3000), verbose = FALSE)

genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
ChIPseeker::vennplot(genes)
```


## 11. Peak overlaps

Without Active Motif
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=8}
peaks = list("Diagenode (1:100)" = GenomicRanges::GRanges(Diagenode_100x),
             "Diagenode (1:50)" = GenomicRanges::GRanges(Diagenode_50x),
             "Abcam 177178 (1:100)" = GenomicRanges::GRanges(Abcam_ab177178),
             "Abcam 4729 (1:100)" = GenomicRanges::GRanges(Abcam_ab4729))

vennplot(peaks)
```


## 12. Motif enrichment

### HOMER

```{r, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=5, fig.height=10}
motifdir = "/Volumes/la420/home/CUTnTAG/antibodies/MACS2/motifs/"
sampleList = c("ActiveMotif", "Diagenode_100x", "Diagenode_50x", "Abcam-ab177178", "Abcam-ab4729")

for(sample in sampleList){
  motifs = read.csv(paste0(motifdir, sample, "/knownResults.txt"), sep = "\t", header = TRUE) %>% data.frame()
  print(head(motifs))
}
```


















