---
title: "CUT&TAG H3K27me3 analysis (no duplicates)"
author: "Leyla Abbasova"
output:
  rmarkdown::html_document:
    theme: default
    highlight: tango
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    number_sections: no
    self_contained: yes
    df_print: paged
---

```{r setup, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(stringr)
library(ggplot2)
library(ggpubr)
library(viridis)
library(GenomicRanges)
library(IRanges)
library(chromVAR)
library(ChIPseeker)
library(ReactomePA)
library(rtracklayer)
library(clusterProfiler)
library(rGADEM)
library(BRGenomics)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
library(GenometriCorr)
```


# General

## 1. Alignment summary 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8, out.width = "100%"}
datadir="/Volumes/la420/home/CUTnTAG/HK5M2BBXY/alignment/sam/bowtie2_summary/"
sampleList = c("H3K27me3")

## collect the alignment results from the bowtie2 alignment summary files
alignResult = c()
for(sample in sampleList){
  alignRes = read.table(paste0(datadir, sample, "_trimmed_bowtie2.txt"), header = FALSE, fill = TRUE)
  alignRate = substr(alignRes$V1[6], 1, nchar(as.character(alignRes$V1[6]))-1)
  alignResult = data.frame(Antibody = sample, SequencingDepth = alignRes$V1[1] %>% as.character %>% as.numeric, 
                           MappedFragNum_hg19 = alignRes$V1[4] %>% as.character %>% as.numeric + alignRes$V1[5] %>% as.character %>% as.numeric,
                           AlignmentRate_hg19 = alignRate %>% as.numeric)  %>% rbind(alignResult, .)
}
alignResult %>% mutate(AlignmentRate_hg19 = paste0(AlignmentRate_hg19, "%"))
```


## 2. Duplicates

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
## summarize the duplication information from the picard summary outputs
picarddir = "/Volumes/la420/home/CUTnTAG/HK5M2BBXY/removeDuplicate/picard_summary/"

dupResult = c()
for(sample in sampleList){
  dupRes = read.table(paste0(picarddir, sample, "_picard.rmDup.txt"), header = TRUE, fill = TRUE)
  dupResult = data.frame(Antibody = sample, MappedFragNum_hg19 = dupRes$READ_PAIRS_EXAMINED[1] %>% as.character %>% as.numeric, DuplicationRate = dupRes$PERCENT_DUPLICATION[1] %>% as.character %>% as.numeric * 100, EstimatedLibrarySize = dupRes$ESTIMATED_LIBRARY_SIZE[1] %>% as.character %>% as.numeric) %>% mutate(UniqueFragNum = MappedFragNum_hg19 * (1-DuplicationRate/100))  %>% rbind(dupResult, .)
}

alignDupSummary = left_join(alignResult, dupResult, by = c("Antibody", "MappedFragNum_hg19")) %>% mutate(DuplicationRate = paste0(DuplicationRate, "%"))
alignDupSummary
```


## 3. Fragment length

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=15, fig.height=8}
## collect the fragment size information
fragdir = "/Volumes/la420/home/CUTnTAG/antibodies_H3K27me3/fragmentLen/"
samples = c("H3K27me3_rmDup", "H3K27me3_withDup")

fragLen = c()
for(sample in samples){
  fragLen = read.table(paste0(fragdir, sample, "_fragmentLen.txt"), header = FALSE) %>% mutate(fragLen = V1 %>% as.numeric, fragCount = V2 %>% as.numeric, Weight = as.numeric(V2)/sum(as.numeric(V2)), Antibody = sample) %>% rbind(fragLen, .) 
}

fig5A = fragLen %>% ggplot(aes(x = Antibody, y = fragLen, weight = Weight, fill = Antibody)) +
    geom_violin(bw = 5) +
    scale_y_continuous(breaks = seq(0, 800, 50)) +
    scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "magma", alpha = 0.8) +
    scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +
    theme_bw(base_size = 20) +
    ggpubr::rotate_x_text(angle = 20) +
    ylab("Fragment Length") +
    xlab("")

fig5B = fragLen %>% ggplot(aes(x = fragLen, y = fragCount, color = Antibody, group = Antibody)) +
  geom_line(size = 1) +
  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = "magma") +
  theme_bw(base_size = 20) +
  xlab("Fragment Length") +
  ylab("Count") +
  coord_cartesian(xlim = c(0, 500))

ggarrange(fig5A, fig5B, ncol = 2)
```


# Peaks

## 1. Import peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir = "/Volumes/la420/home/CUTnTAG/antibodies_H3K27me3/peaks"

H3K27me3_SEACR_path = file.path(peakdir, "H3K27me3_SEACR_peaks.bed")
H3K27me3_SEACR = ChIPseeker::readPeakFile(H3K27me3_SEACR_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)

H3K27me3_MACS2_path = file.path(peakdir, "H3K27me3_MACS2_peaks.bed")
H3K27me3_MACS2 = ChIPseeker::readPeakFile(H3K27me3_MACS2_path, as = "GRanges") %>% BRGenomics::tidyChromosomes(keep.X = TRUE, keep.Y = TRUE)
```


## 2. Peak information

### Peak numbers

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakdir = "/Volumes/la420/home/CUTnTAG/antibodies_H3K27me3/peaks/"
sampleList = c("H3K27me3_SEACR", "H3K27me3_MACS2")

Total_CT_peaks = c()
for(sample in sampleList){
    peakInfo_S = read.table(paste0(peakdir, sample, "_peaks.bed"), header = FALSE, fill = TRUE)
    Total_CT_peaks = data.frame(Total_CT_peaks = nrow(peakInfo_S), Antibody = sample) %>% rbind(Total_CT_peaks, .)
}

Total_CT_peaks %>% dplyr::select(Antibody, Total_CT_peaks)
```

### Peak widths

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
samples = list(H3K27me3_SEACR, H3K27me3_MACS2)

for(sample in samples){
  print(GenomicRanges::width(sample) %>% summary())
}
```


### Fraction of reads in peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
datadir = "/Volumes/la420/home/CUTnTAG/antibodies_H3K27me3/peaks/"
bamdir = "/Volumes/la420/home/CUTnTAG/HK5M2BBXY/alignment/bam/"
sampleList = c("H3K27me3_SEACR", "H3K27me3_MACS2")

inPeakData = c()
for(sample in sampleList){
  peakRes = read.table(paste0(datadir, sample, "_peaks.bed"), header = FALSE, fill = TRUE)
  peak.gr = GenomicRanges::GRanges(seqnames = peakRes$V1, IRanges(start = peakRes$V2, end = peakRes$V3), strand = "*")
  bamFile = paste0(bamdir, "H3K27me3_bowtie2_rmDup.mapped.bam")
  fragment_counts = getCounts(bamFile, peak.gr, paired = TRUE, by_rg = FALSE, format = "bam")
  inPeakN = counts(fragment_counts)[,1] %>% sum
  inPeakData = rbind(inPeakData, data.frame(inPeakN = inPeakN, Antibody = sample))
}

frip = left_join(alignDupSummary, inPeakData, by = "Antibody") %>% mutate(frip = inPeakN/UniqueFragNum * 100)
frip %>% dplyr::select(Antibody, SequencingDepth, MappedFragNum_hg19, AlignmentRate_hg19, UniqueFragNum, FragInPeakNum = inPeakN, FRiPs = frip)
```


## 3. Coverage   {.tabset}

### SEACR

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
H3K27me3_SEACR_adj = data.frame(H3K27me3_SEACR)
# remove 30 largest peaks
H3K27me3_SEACR_adj = H3K27me3_SEACR_adj[H3K27me3_SEACR_adj$V4 < 400000, ] %>% GenomicRanges::GRanges()

ChIPseeker::covplot(peak = H3K27me3_SEACR_adj,
                    weightCol = "V4",
                    title = "CUT&TAG H3K27me3 peaks (SEACR)")
```

### MACS2

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=8, fig.height=10}
H3K27me3_MACS2_adj = data.frame(H3K27me3_MACS2)
# remove 10 largest peaks
H3K27me3_MACS2_adj = H3K27me3_MACS2_adj[H3K27me3_MACS2_adj$V5 < 600, ] %>% GenomicRanges::GRanges()

ChIPseeker::covplot(peak = H3K27me3_MACS2_adj,
                    weightCol = "V5",
                    title = "CUT&TAG H3K27me3 peaks (MACS2)")
```


## 4. ENCODE peak overlap

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
extraCols_narrowPeak = c(signalValue = "numeric", pValue = "numeric", qValue = "numeric", peak = "integer")

ENCODE_narrowPeaks = rtracklayer::import("https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz", format = "BED", extraCols = extraCols_narrowPeak)

ENCODE_broadPeaks = rtracklayer::import("https://www.encodeproject.org/files/ENCFF001SZE/@@download/ENCFF001SZE.bed.gz", format = "BED", extraCols = extraCols_narrowPeak)
```


### Count total C&T peaks falling into ENCODE (narrow peaks)
Discard values of 0 (corresponding to peak ranges with 0 ENCODE overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peak_list = list(H3K27me3_SEACR, H3K27me3_MACS2)
total_overlapping_peaks_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = peak_list[[i]], subject = ENCODE_narrowPeaks)
  CT_peaks_in_ENCODE = overlaps[overlaps != 0] %>% length()
  total_overlapping_peaks_list = c(total_overlapping_peaks_list, CT_peaks_in_ENCODE)
}

Total_CT_peaks_falling_into_ENCODE = data.frame(Antibody = sampleList, Total_CT_peaks_falling_into_ENCODE = total_overlapping_peaks_list)
Total_CT_peaks_falling_into_ENCODE
```

### Count total ENCODE peaks falling into C&T
Discard values of 0 (corresponding to peak ranges with 0 C&T overlaps) and count how many ranges (peaks) remain:
```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
captured_ENCODE_list = c()

for(i in 1:length(peak_list)){
  overlaps = GenomicRanges::countOverlaps(query = ENCODE_narrowPeaks, subject = peak_list[[i]])
  ENCODE_captured_peaks = overlaps[overlaps != 0] %>% length()
  captured_ENCODE_list = c(captured_ENCODE_list, ENCODE_captured_peaks)
}

Total_captured_ENCODE_peaks = data.frame(Antibody = sampleList, Total_captured_ENCODE_peaks = captured_ENCODE_list)
Total_captured_ENCODE_peaks
```


### Summarize

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ENCODE_overlaps = Total_captured_ENCODE_peaks %>% left_join(Total_CT_peaks, by = "Antibody") %>% left_join(Total_CT_peaks_falling_into_ENCODE, by = "Antibody")

ENCODE_overlaps$Percentage_of_CT_peaks_falling_into_ENCODE = ENCODE_overlaps$Total_CT_peaks_falling_into_ENCODE / ENCODE_overlaps$Total_CT_peaks * 100
ENCODE_overlaps$Percentage_of_total_ENCODE = ENCODE_overlaps$Total_captured_ENCODE_peaks / length(ENCODE_narrowPeaks) * 100

ENCODE_overlaps = ENCODE_overlaps[c("Antibody", "Total_CT_peaks", "Total_CT_peaks_falling_into_ENCODE", "Percentage_of_CT_peaks_falling_into_ENCODE", "Total_captured_ENCODE_peaks", "Percentage_of_total_ENCODE")]
ENCODE_overlaps
```


## 5. SEACR / MACS2 peak overlap

### Count total MACS2 peaks overlapping with SEACR peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
overlaps = GenomicRanges::countOverlaps(query = H3K27me3_MACS2, subject = H3K27me3_SEACR)
MACS2_in_SEACR = overlaps[overlaps != 0] %>% length()

Total_MACS2_peaks = length(H3K27me3_MACS2)
MACS2_in_SEACR = data.frame(Antibody = "H3K27me3_MACS2", Total_MACS2_peaks = Total_MACS2_peaks, Total_MACS2_peaks_in_SEACR = MACS2_in_SEACR)
MACS2_in_SEACR$Percentage_of_total_MACS2 = MACS2_in_SEACR$Total_MACS2_peaks_in_SEACR/MACS2_in_SEACR$Total_MACS2_peaks*100

MACS2_in_SEACR
```


### Count total SEACR peaks overlapping with MACS2 peaks

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
overlaps = GenomicRanges::countOverlaps(query = H3K27me3_SEACR, subject = H3K27me3_MACS2)
SEACR_in_MACS2 = overlaps[overlaps != 0] %>% length()

Total_SEACR_peaks = length(H3K27me3_SEACR)
SEACR_in_MACS2 = data.frame(Antibody = "H3K27me3_SEACR", Total_SEACR_peaks = Total_SEACR_peaks, Total_SEACR_peaks_in_MACS2 = SEACR_in_MACS2)
SEACR_in_MACS2$Percentage_of_total_SEACR = SEACR_in_MACS2$Total_SEACR_peaks_in_MACS2/SEACR_in_MACS2$Total_SEACR_peaks*100

SEACR_in_MACS2
```


## 6. Average TSS peak profiles

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
promoters = ChIPseeker::getPromoters(TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, upstream = 3000, downstream = 3000)
peak_files = list("H3K27me3 (SEACR)" = H3K27me3_SEACR,
                  "H3K27me3 (MACS2)" = H3K27me3_MACS2,
                  "ENCODE narrow peaks" = ENCODE_narrowPeaks)

tagMatrixList = lapply(peak_files, getTagMatrix, windows = promoters)

ChIPseeker::plotAvgProf(tagMatrixList, xlim = c(-3000, 3000), conf = 0.95, resample = 500, facet = "row")
```


## 7. Peak heatmaps

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
tagHeatmap(tagMatrixList, xlim = c(-3000, 3000), color = NULL)
```


## 8. Annotation

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
peakAnnoList = lapply(peak_files, annotatePeak, TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene, tssRegion = c(-3000, 3000), verbose = FALSE)

ChIPseeker::plotAnnoBar(peakAnnoList)
```


## 9. Distance to TSS

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
ChIPseeker::plotDistToTSS(peakAnnoList)
```


## 10. Enrichment analysis

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=13, fig.height=8}
genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))
compKEGG = clusterProfiler::compareCluster(geneCluster = genes,
                                           fun = "enrichKEGG",
                                           pvalueCutoff = 0.05,
                                           pAdjustMethod = "BH")
clusterProfiler::dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


## 11. Motif enrichment

### HOMER

```{r, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE, fig.width=5, fig.height=10}
motifdir = "/Volumes/la420/home/CUTnTAG/antibodies_H3K27me3/motifs/"
sampleList = c("H3K27me3_SEACR", "H3K27me3_MACS2")

for(sample in sampleList){
  motifs = read.csv(paste0(motifdir, sample, "_knownResults.txt"), sep = "\t", header = TRUE) %>% data.frame()
  print(head(motifs))
}
```








