---
title: "nf-core_atacseq_results"
author: "Brian M. Schilder"
date: "Most recent update:<br> `r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: hide 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true  
---

```{r setup}
library(PeakyFinders)
library(EpiCompare)
```


# Gather files

## Native CHiP-seq

[GSE66023](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE66023&s=09)

Specifically, the peak files for K562 cells are found [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2054696). 

All data was aligned to either mm9 or hg19 in the [original study](https://www.nature.com/articles/ng.3550).

```{r} 
grl_native <- PeakyFinders::import_peaks(ids = "GSM2054696",
                                         builds = "hg19",
                                         searches = list(genericPeak=".bed.gz"))
unique(grl_native$GEO[[1]]$source)

peaks_native <- grl_native$GEO[[1]]

peaks_native$qValue <- peaks_native$V7
peaks_native <- list(
  "GSM2054696_K562_H3K27ac"=peaks_native[grepl("H3K27ac",peaks_native$source), ]
)
```

Data was reprocessed from fastsq by Leyla Abbasova to make consistent with our CUT&Tag analysis pipeline. All data was aligned to hg19.

```{r}

peaks_native <- rtracklayer::import("/Volumes/bms20/projects/neurogenomics-lab/live/Projects/CUT_n_TAG/native_ChIP/K562_H3K27ac_native_ChIP_SRR3144862.srt.nodup_x_K562_input_native_ChIP_SRR3144864.srt.nodup.pval0.01.500K.bfilt.narrowPeak.gz",
                                    format="narrowPeak")
peaks_native <- list("GSM2054696_K562_H3K27ac"=peaks_native) 
```


## ENCODE

After confirming on the ENCODE website, ENCODE does not contain any data for
the marks H3K122 and H3K64ac (only H3K27ac). 

```{r} 
marks <- c("H3K122","H3K27ac","H3K64ac","H3K27me3")
meta <- PeakyFinders::search_encode(target =  c("H3K122","H3K27ac",
                                                "H3K64ac","H3K27me3"),
                      biosample_name = "K562",
                      organism = "Homo sapiens",
                      assembly = "GRCh38",
                      file_type = "narrowPeak",
                      output_type=c("replicated peaks"), # "^peaks$"
                      peaks_only = TRUE) 
meta <- meta[grepl(paste(marks,collapse = "|"), target, ignore.case = TRUE),]

reference <- PeakyFinders::import_peaks(ids = meta$file_accession,  # "ENCSR000AKP" 
                                        builds = "GRCh38")
```

## CUT&Tag

Our data from the CUT&Tag preprint: 
[GSE199611](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE199611)

```{r}
#### Import peak files from GEO ####
peaks_ct <- PeakyFinders::import_peaks(
  ids = "GSE199611",
  builds = "hg19",
  searches = list(genericPeak=".bed.gz"))
#### Name peak files #####
peaks_ct2 <- peaks_ct$GEO
names(peaks_ct2) <- gsub(".rmDup|.bed|.gz", "",unlist(lapply(peaks_ct2, function(x){x$source[1]})))
names(peaks_ct2) <- paste(
  "C&T",
  ifelse(grepl("CST9733",names(peaks_ct2)),"H3k27me3","H3K27ac"),
  names(peaks_ct2),
  sep = ".")

#### Create peak file groups #####
peaks_grouped <- EpiCompare::group_files(
  peakfiles = peaks_ct2, 
  searches = list(assay="CUT&Tag",
                  histone=c("H3k27me3","H3K27ac"),
                  peak_caller=c("MACS2","SEACR")))

consensus_peaks_ct <- EpiCompare::compute_consensus_peaks(
  grlist = peaks_ct2, 
  groups = peaks_grouped[names(peaks_ct2)], 
  genome_build = "hg19", 
  method = "consensusseeker",
  nbrThreads = 10)
# saveRDS(consensus_peaks_ct,"consensus_peaks.GSE199611.rds")
# consensus_peaks_ct <- readRDS("consensus_peaks.GSE199611.rds")
consensus_peaks_ct <- EpiCompare::liftover_grlist(grlist = consensus_peaks_ct,
                                         input_build = "hg19",
                                         output_build = "hg38")
```

## TIP-seq

```{r}
peaks_tip <- readRDS("~/Downloads/peaks_tip_hg38.rds")
names(peaks_tip) <- paste("TIP-seq",names(peaks_tip),sep=".")

peaks_grouped_tip <- EpiCompare::group_files(
  peakfiles = peaks_tip,
  searches = list(assay=c("TIP-seq"),
                  histone=c("H3k27me3","H3K27ac")) 
  )
consensus_peaks_tip <- EpiCompare::compute_consensus_peaks(
  grlist = peaks_tip, 
  groups = peaks_grouped_tip[names(peaks_tip)], 
  genome_build = "hg38", 
  method = "consensusseeker",
  nbrThreads = 10)
saveRDS(consensus_peaks_tip,"consensus_peaks_tip.rds")
consensus_peaks_tip <- readRDS("consensus_peaks_tip.rds")
```

# Run EpiCompare

## CUT&Tag / TIP-seq vs. ENCODE

```{r} 
data("hg38_blacklist") 

peakfiles_all <- c(consensus_peaks_ct,consensus_peaks_tip)
peakfiles_all <- peakfiles_all[sort(names(peakfiles_all))]

res <- EpiCompare::EpiCompare(peakfiles = peakfiles_all,
                              genome_build = list(peakfiles="hg38",
                                                  reference="hg38",
                                                  blacklist="hg38"), 
                              blacklist = hg38_blacklist,
                              genome_build_output = "hg38",
                              reference = list(ENCODE_H3K27ac_ENCFF038DDS=reference$ENCODE$ENCFF038DDS,
                                               ENCODE_H3K27me3_ENCFF049HUP=reference$ENCODE$ENCFF049HUP),
                              chromHMM_plot = TRUE,
                              chromHMM_annotation = "K562",
                              chipseeker_plot = TRUE,
                              enrichment_plot = TRUE,
                              tss_plot = TRUE, 
                              stat_plot = TRUE,
                              upset_plot = TRUE,
                              # precision_recall_plot = TRUE,
                              # corr_plot = TRUE,
                              interact = TRUE,
                              save_output = TRUE,
                              output_dir = here::here("EpiCompare_CnT_TIP_consensus"))
```   

### Precision-recall curves

#### H3K27ac

```{r}

peakfiles_ac <- peakfiles_all[grepl("H3K27ac",names(peakfiles_all), ignore.case = TRUE)]
pr_ac <- EpiCompare::plot_precision_recall(
  peakfiles = peakfiles_ac,
  reference = 
    list(ENCODE_H3K27ac_ENCFF038DDS=reference$ENCODE$ENCFF038DDS),
  subtitle = "H3K27ac"
  ) 
```

#### H3K27me3

```{r}
peakfiles_me3 <- peakfiles_all[grepl("H3K27me3",names(peakfiles_all), ignore.case = TRUE)]
pr_me3 <- EpiCompare::plot_precision_recall(
  peakfiles = peakfiles_me3,
  reference =
    list(ENCODE_H3K27me3_ENCFF049HUP=reference$ENCODE$ENCFF049HUP),
 subtitle = "H3K27me3") 
```
## Native CHiP-seq vs. ENCODE

```{r} 
data("hg38_blacklist")  
res <- EpiCompare::EpiCompare(peakfiles = peaks_native,
                              genome_build = list(peakfiles="hg19",
                                                  reference="hg38",
                                                  blacklist="hg38"), 
                              blacklist = hg38_blacklist,
                              genome_build_output = "hg38",
                              reference = list(ENCODE_H3K27ac_ENCFF038DDS=reference$ENCODE$ENCFF038DDS),
                              chromHMM_plot = TRUE,
                              chromHMM_annotation = "K562",
                              chipseeker_plot = TRUE,
                              enrichment_plot = TRUE,
                              tss_plot = TRUE, 
                              stat_plot = TRUE,
                              upset_plot = TRUE,
                              precision_recall_plot = TRUE,
                              corr_plot = TRUE,
                              interact = TRUE,
                              save_output = TRUE,
                              output_dir = here::here("EpiCompare_nativeCHiPseq"))
```   

# Session info

<details>

```{r}
utils::sessionInfo()
```

</details>

<br>
