---
title: "visualize_peaks"
author: "Brian M. Schilder"
date: "Most recent update:<br> `r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true  
---
```{r setup, include=T, message=F, dpi=300,  root.dir="/rds/general/project/neurogenomics-lab/live/GitRepos/CUT_n_TAG"}
# root.dir <- "/rds/general/project/neurogenomics-lab/live/GitRepos/CUT_n_TAG"
# root.dir <- "/Volumes/RDS/project/neurogenomics-lab/live/GitRepos/CUT_n_TAG" 
root.dir <- "/Volumes/bms20/projects/neurogenomics-lab/live/GitRepos/CUT_n_TAG"

CT.dir <- file.path(root.dir,"processed_data/HK5M2BBXY_merged")
CT.peaks_dir <- file.path(CT.dir,"bwa/mergedLibrary/macs/narrowPeak")
CT.bw_dir <- file.path(CT.dir,"bwa/mergedLibrary/bigwig")

source("functions.R")
try({setwd(root.dir)})
knitr::opts_chunk$set(echo = T, root.dir = root.dir)
knitr::opts_knit$set(root.dir = root.dir)

library(dplyr) 
# library(echolocatoR) # devtools::install_github("RajLabMSSM/echolocatoR")
library(rtracklayer) # BiocManager::install("rtracklayer")
library(ggbio) # BiocManager::install("ggbio")
library(rGADEM) # BiocManager::install("rGADEM")
library(BSgenome.Hsapiens.UCSC.hg19) #BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
library(ChIPseeker) #BiocManager::install("ChIPseeker")
library(ggupset) # install.packages("ggupset")
library(ggimage) # install.packages("ggimage")
library(clusterProfiler)
library(ReactomePA)  # BiocManager::install("ReactomePA")

library(rGADEM)
library(BSgenome.Hsapiens.UCSC.hg19)
library(GenomicRanges)

# IMPORTANT! Otherwise can have issues with rtracklayer::import()
# base::closeAllConnections()
```



 The following scripts primarily follows these tutorials:
 - [genomation]()  
 - [ChIPseeker](http://bioconductor.org/packages/devel/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html)  


# Data descriptions  

## ENCODE  

All ENCODE peak/bigwig/bam/fastq.gz files can be found [here on UCSC](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHistone/). 


For general info on file formats (BED, narrowPeak, broadPeak) see [UCSC Genome Browser documentation](https://genome.ucsc.edu/FAQ/FAQformat.html). Specifically, there are [3 ENCODE histone datasets](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/):  

* [wgEncode Broad Histone](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHistone/)
  - *broadPeak*: Yes  
  - *narrowPeak*: No  
  - *H3k27ac in K562 cell-line*: Yes.  
  - *H3k27me3 in K562 cell-line*: Yes.  

* [wgEncode Uw Histone](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwHistone/)  
  - *broadPeak*: Yes    
  - *narrowPeak*: Yes  
  - *H3k27ac in K562 cell-line*: No.  
  - *H3k27me3 in K562 cell-line*: Yes.  
  
- [wgEncode Sydh Histone](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeSydhHistone/)  
  - *broadPeak*: No  
  - *narrowPeak*: Yes  
  
Peak files are also available for all three datasets in [.bb format on UCSC](ttp://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/jan2011/histone_macs/optimal/hub/).
  - The latest version of `rtracklayer` (>=1.5) has `import.bb` function, but can't seem to install it...   
  
**TF-specific** ChIP-seq ENCODE (narrow) peak files can again be found [here](http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/june2012/spp/optimal/). 

  
Several other datasets of potential interest for comparisons to CUT&TAG (though they only contain bigWig files, not peak files):  

* wgEncodeRegMarkH3k4me1/  
* wgEncodeRegMarkH3k4me3/  
* wgEncodeRegMarkH3k27ac/  

### ENCODE metadata  

```{r ENCODE metadata, eval=F}
# contains links and metadata for all ENCODE files, but the format is not very readable...
meta <- data.table::fread("http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/files.txt", sep=";", sep2 = " ") 
links <- meta$V1 # links to .bb files?
```
 
# Create query data  

Import a `GRanges` object from the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny/) to use for querying a small subset of the CUT&TAG data.  

```{r Create query data, eval=F}
file_names <- echolocatoR::GITHUB.list_files(creator = "RajLabMSSM", 
                                repo = "Fine_Mapping_Shiny",
                                query = "*Nalls23andMe_2019.*BST1.UKB.multi_finemap.csv.gz")
gr.dat <- GenomicRanges::makeGRangesFromDataFrame(data.table::fread(file_names), 
                                                  keep.extra.columns = T, 
                                                  seqnames.field = "CHR", 
                                                  start.field = "POS",
                                                  end.field = "POS")
# ! IMPORTANT !: Needs to be in same chromosome format as bigwig in order to query!
suppressWarnings(GenomeInfoDb::seqlevelsStyle(gr.dat) <- "UCSC")
```

# ENCODE  

## .broadPeak  

* Naming conventions for broadPeak come from [UCSC](https://genome.ucsc.edu/FAQ/FAQformat.html#format13). 
* IMPORTANT!: Note that "pValue" and "qValue" are actually the -log() of these metrics. Additionally, -1 means that this metric is not available (instead of NA?....)

```{r Import peaks}  
ENCODE.broadPeaks <- rtracklayer::import("http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHistone/wgEncodeBroadHistoneK562H3k27acStdPk.broadPeak.gz")  
```

## .bigWig

* ENCODE bigWig and peaks data is hosted on the [UCSC Genome Browser here](http://genome.ucsc.edu/cgi-bin/hgTracks?tsCurTab=advancedTab&tsGroup=Any&tsType=Any&hgt_mdbVar1=dccAccession&hgt_tSearch=search&hgt_tsDelRow=&hgt_tsAddRow=&hgt_tsPage=&tsSimple=&tsName=&tsDescr=&db=hg19&hgt_mdbVal1=wgEncodeEH000043). 
* Specifcally, we'll be querying the H3K27ac ChIP-seq data from K562 cell-lines, to best match up with the CUT&TAG assay being used here. 

```{r Query ENCODE bigwig, eval=F}
ENCODE.bw_filt <- import.bw.parallel(bw.file = "http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHistone/wgEncodeBroadHistoneK562H3k27acStdSig.bigWig",
                                    gr.dat = ENCODE.broadPeaks, 
                                    bw.file_format = "UCSC")
head(ENCODE.broadPeaks)
```

# CUT&TAG  

* When I ran the nf-core/atacseq pipeline for the first time on this CUT&TAG data, I accidentally mis-specified the design matrix such that the H3k27ac and H3k27ame3 assays were merged into one (in the *mergedReplicates* subfolder).  

* However, we can still recover the independent assays from the *mergedLibrary* since each assay only had one sample. In this case, use these keys:
  - control_R1 = H3k27ac  
  - control_R2 = H3k27ame3  
  
**File types**:  
The nf-core/atacseq pipeline produces of number of peaks-related files.  
- bigwig  
- summits  
- annotatePeaks  
- narrowPeaks and/or broadPeaks  

## .bigWig

* Import a subset of a bigWig file based on the coordinates in the `GRanges` object (ENCODE broadPeaks).  

```{r .bigwig, eval=F} 
CT.bw_filt <- import.bw.parallel(bw.file = file.path(CT.bw_dir,
                                                     "control_R1.mLb.clN.bigWig"),
                          gr.dat = ENCODE.broadPeaks, 
                          bw.file_format = "NCBI")
head(CT.bw_filt)
```

## .summits

Summits are especially the peaks of the peaks (1bp/peak). 

```{r .summits}
CT.summits <- rtracklayer::import(file.path(CT.peaks_dir,
                                            "control_R1.mLb.clN_summits.bed")) 
CT.summits$width <- GenomicRanges::width(CT.summits)
summary(CT.summits$width)
```


## .annotatePeaks  

This previously annotated file contains additional information like which gene each peak is closest to.  

```{r .annotatePeaks} 
# Narrow peaks
CT.annotatePeaks <-  data.table::fread(file.path(CT.peaks_dir,"control_R1.mLb.clN_peaks.annotatePeaks.txt")) %>%
  dplyr::mutate(peak_score=`Peak Score`) %>%
  GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "Chr", 
                                          start.field = "Start", end.field = "End",strand.field = "Strand",
                                          keep.extra.columns = T) 
CT.annotatePeaks$width <- GenomicRanges::width(CT.annotatePeaks) 
summary(CT.annotatePeaks$width) 
```

## .narrowPeak  

* Column names derived from [UCSC documentation](https://genome.ucsc.edu/FAQ/FAQformat.html#format12).  

* Interestingly, these peaks are slightly different lengths than those in *.annotatePeaks*.  

* *NOTE!*:`rtracklayer::import` will get confused by a narrowPeak file that is missing the extra stats columns, which are missing here because it's not a *mergedReplicates* file. Instead, just read it in as a regular bed file for now.

```{r .narrowPeak}
CT.narrowPeaks <- rtracklayer::import.bed(file.path(CT.peaks_dir,
                                                "control_R1.mLb.clN_peaks.narrowPeak")) 
CT.narrowPeaks$width <- GenomicRanges::width(CT.narrowPeaks)  
summary(CT.narrowPeaks$width)

ChIPseeker::covplot(peak = CT.narrowPeaks,
                    weightCol = "score",#"qValue",
                    title = "CUT&TAG peaks")
```



# CUT&TAG vs. ENCODE  

## Prepare peaks lists  

```{r Prepare peaks lists}
suppressWarnings(GenomeInfoDb::seqlevelsStyle(CT.narrowPeaks) <- "UCSC")
peaks_list <- list("ICL CUT&TAG"=CT.narrowPeaks, 
                   "ENCODE ChIP-seq"=ENCODE.broadPeaks)

tagMatrix_list <- prepare_tagMatrix(peaks_list=peaks_list)
annotatePeak_list <- prepare_annotatePeak(peaks_list = peaks_list)
```

## Compare overlap  

```{r Compare overlap}
compare_peak_overlap(gr.query = CT.narrowPeaks, 
                     gr.subject = ENCODE.broadPeaks)
```
## Compare peaks

Compare peaks and their annotations across datasets using tools like [`ChIPseeker`](http://bioconductor.org/packages/devel/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html). 

### tagHeatmap

```{r tagHeatmap} 
ChIPseeker::tagHeatmap(tagMatrix_list, 
                       xlim=c(-3000, 3000))
```

### plotAvgProf

```{r plotAvgProf}
ChIPseeker::plotAvgProf(tagMatrix_list,
                        conf = 0.95, resample = 100,
                        xlim=c(-3000, 3000), 
                        facet="row")
```

### plotAnnoBar

```{r plotAnnoBar}
ChIPseeker::plotAnnoBar(annotatePeak_list)
```

### plotDistToTSS  

```{r plotDistToTSS}
ChIPseeker::plotDistToTSS(annotatePeak_list)
```
 
### upsetplot

```{r upsetplot}
# peakAnno <- annotatePeak_list$CT
# ChIPseeker::plotAnnoPie(peakAnno)
# ChIPseeker::plotAnnoBar(peakAnno)
# ChIPseeker::vennpie(peakAnno)
lapply(annotatePeak_list, function(x){ChIPseeker::upsetplot(x, vennpie=T)})
```
## Gene enrichment  

There are different strategies for assigning each peak to a gene. ChIPseeker provides two main alternatives:  
1. `annotatePeak()`: Default method, used by `prepare_annotatePeak()`. 
2. `seq2gene()`: Alternative method, can be used with `use_seq2gene=T` argument where applicable.  

### clusterProfiler  

Use [clusterProfiler](http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) to compare functional profiles across datasets.  

```{r Functional profiles comparison}
res_clusterProfiler <- enrich_clusterProfiler(annotatePeak_list,
                                              fun="enrichKEGG",
                                              pvalueCutoff  = 0.05,
                                              pAdjustMethod = "BH", 
                                              use_seq2gene = F, 
                                              show_plot = T)
```

### ReactomePA enrichment  

```{r ReactomePA}
res_ReactomePA <- enrich_ReactomePA(annotatePeak_list,
                                    pvalueCutoff  = 0.05,
                                    pAdjustMethod = "BH", 
                                    use_seq2gene = T)
```

### Gene overlap  

```{r Overlap of peaks and annotated genes}
vp_res <- gene_vennplot(annotatePeak_list = annotatePeak_list, 
                        use_seq2gene = F)

vp_res.seq2gene <- gene_vennplot(annotatePeak_list = annotatePeak_list, 
                        use_seq2gene = T)
```

## Statistical testing of peak overlap

* Calculate whether overlap is statistically significant between two datasets of peaks.  
* In this case, we're comparing CUT&TAG peaks to ENCODE ChiP-seq peaks.

### Convert to UCSC format  

```{r Convert to UCSC format}
CT.UCSC_path <- file.path(CT.peaks_dir,"control_R1.mLb.clN_peaks.narrowPeak")
rtracklayer::export.bed(CT.narrowPeaks, 
                        con = CT.UCSC_path)
```


* `enrichPeakOverlap` iterates enrichment tests over many perturbations (shuffle), parallizing across `detectCores() - 1` cores.
* Parameter `targetPeak` can be the folder, e.g. hg19, that containing bed files. 

```{r enrichPeakOverlap}
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene 

peak_enrich <- ChIPseeker::enrichPeakOverlap(queryPeak = CT.UCSC_path,
                                 targetPeak    = file.path(root.dir,"raw_data/ENCODE","wgEncodeBroadHistoneK562H3k27acStdPk.broadPeak.gz"),
                                 TxDb          = txdb,
                                 pAdjustMethod = "BH",
                                 nShuffle      = 100,
                                 chainFile     = NULL,
                                 verbose       = T)
print(peak_enrich)
```

# Discover TFBS motifs  

*under construction*  

* From the genomatic [documentation](https://compgenomr.github.io/book/motif-discovery.html#motif-comparison). 
* Motif discovery step uses [`rGADEM` R package](https://bioconductor.org/packages/release/bioc/vignettes/rGADEM/inst/doc/rGADEM.pdf). 

* [MotifbreakR](http://bioconductor.org/packages/release/bioc/html/motifbreakR.html) could also be used, but is more designed for individual SNPs rather than genomic ranges (to know knowledge). See `echolocatoR::MOTIFBREAKR()` for a convenience wrapper of this package.

```{r Discover CUT&TAG motifs, eval=F}
# order the peaks by qvalue, and take top 250 peaks
CT.narrowPeaks_top = CT.narrowPeaks[order(CT.narrowPeaks$pValue)]
CT.narrowPeaks_top = head(CT.narrowPeaks_top, n = 500)
# CT.narrowPeaks_top_stored <- CT.narrowPeaks_top
# merge nearby  peaks
CT.narrowPeaks_top = GenomicRanges::reduce(CT.narrowPeaks_top, )
# Create a region of  +/-50 bp around the center of the peaks
CT.narrowPeaks_top = GenomicRanges::resize(CT.narrowPeaks_top, 
                                           width = 50, fix='center')
suppressWarnings(GenomeInfoDb::seqlevelsStyle(CT.narrowPeaks_top) <- "UCSC")
CT.narrowPeaks_top$score <- GenomicRanges::score(CT.narrowPeaks_top)
CT.narrowPeaks_top$width <- GenomicRanges::width(CT.narrowPeaks_top)

## Example data
# pwd<-"" #INPUT FILES- BedFiles, FASTA, etc.
# path<- system.file("extdata/Test_100.bed",package="rGADEM")
# BedFile<-paste(pwd,path,sep="")
# Sequences <- rtracklayer::import(BedFile)
# Sequences <- rtracklayer::import(BedFile)

gadem <- GADEM(Sequences=CT.narrowPeaks_top,
               verbose=1,
               genome=Hsapiens)
```


# Session Info

<details>

```{r Session Info}
sessioninfo::session_info()
```

</details>

